accepted/wasmtime-one-dot-oh.md

```diff
@@ -0,0 +1,558 @@
+
+* If the feature is for a WebAssembly upstream proposal, all spec tests must be
+  enabled and passing and the proposal must be in [stage 4 or
+  later](https://github.com/WebAssembly/meetings/blob/main/process/phases.md).
+
+* The feature must have no open questions, design concerns, or serious known
+  bugs.
+
+* The feature must have support in fuzzers and have been fuzzed for at least a
+  week. Additionally we should be confident that the fuzzing support for this
+  feature is exercising the necessary bits in Wasmtime thoroughly.
+
+* The feature is supported in the Rust API, the C API, and at least one other
+  language's embedding API.
+
+* A member of the Bytecode Alliance must be "on the hook" for maintenance of
+  this feature.
+
+Features can be implemented in-tree even if they do not meet these criteria but
+the features must be disabled-by-default at either runtime or compile
+time. If a feature's interim implementation does not have an undue compile-time
+or runtime footprint then it can be off-by-default at runtime but compiled in by
+default. If, however, an in-progress feature has a significant compile-time or
+runtime footprint it must be disabled at compile-time by default.
+
+Features implemented in-tree but not currently stabilized are also subject to
+removal if there is no active progress being made on the feature. For example if
+this RFC is approved **the `lightbeam` feature of Wasmtime would be removed**.
+
+> **Note**: This RFC proposes removing `lightbeam` because it isn't actively
+> maintained, and as a result hasn't been compiling successfully for quite
+> some time. The Wasmtime project is still interested in the addition of a
+> baseline compiler, but the addition of one would be proposed in an RFC and
+> undergo careful evaluation to ensure the design aligns with Wasmtime's
+> goals and requirements, and the compiler has a solid maintenance story.
+
+## Release Process and Cadence
+
+Wasmtime intends to follow in the footsteps of many other projects on the matter
+of release cadence with a frequent and predictable release cycle. **This RFC
+proposes releasing Wasmtime on Tuesday every 4 weeks**. The precise date of
+each release may be adjusted to avoid coinciding with public holidays, though,
+which could result in some releases being a slightly different width apart.
+
+The goal is to publish work and improvements in Wasmtime on a relatively rapid
+schedule to ensure that the latest-and-greatest is available for usage. This is
+also intended to signal to users that consumers of Wasmtime should expect to
+stay up-to-date with Wasmtime and it's not appropriate to ossify a version for
+years and expect it to continue to work and receive everlasting maintenance.
+
+Each release of Wasmtime will bump the major version number of Wasmtime itself.
+This means, for example, that Wasmtime 2.0 will be released one month after
+Wasmtime 1.0. After one year Wasmtime will be at 12.0. At this time it is not
```

Comment:
I would not expect this as a user: I would expect the project to follow semantic versioning and decide every four weeks if the release should be major or minor. Users then get some type of clue about the changes coming to them ("oh, this is a major change, some API must have changed--let me go look at the release notes!"). It seems that the "match the Wasmtime version with the bindings version" can be solved some other way (e.g., `wasmtime-py [Wasmtime version]-[binding version]`)

Reply:
That may have been a usual expectation at some point in the past, but many projects have changed their versioning policies in recent years to this sort of release train, for example the LLVM project switched to this in 2016 http://blog.llvm.org/2016/12/llvms-new-versioning-scheme.html

Reply:
Most of these projects do have breaking changes every release now. Wasmtime probably won't have breaking changws every time when releasing at the suggested interval.

Reply:
I started out with this view, too, but talking through it with Alex and others made me pretty convinced that proper fine-grained semver handling would be very, perhaps prohibitively, hard.

As described in the explainer, the language embeddings are a major consideration here: I think it's very clearly desirable to keep their version numbers aligned with Wasmtime's (because you shouldn't have to maintain a mapping table from language embedding version to Wasmtime version in your head). I really don't see a way to achieve this without regularly scheduled version bumps, in particular if we want to enable third-party embeddings to stay version aligned, too.

Reply:
If we bump the major version of all crates part of wasmtime to the same version even if only some crates have breaking changes, I guess it wouldn't be much of a stretch to bump the major version every time any language embedding has a breaking change.

Reply:
That's true, yes. But I think it'd in reality have the effect of making it harder for embeddings to improve in ways that require these kinds of changes. In particular for lower-tier ones.

As Pat mentioned above, we're not proposing to tread particularly new ground here: LLVM does pretty much the same, and I'd assume for much the same reasons.

Reply:
Personally my experience is that maintaining a strategy where you figure out whether breaking changes or not have happened is onerous and difficult to maintain. From the perspective of authoring Wasmtime it means that we'll quickly fall into a trap of "oh when's the right time to have a breaking change?" and try to batch things up, in my opinion creating unnecessary tension for landing improvements and updates to the engine.

From a user's perspective I would consider Wasmtime as a "weighty" enough dependency that even if you want to update patch versions there's likely to be some sort of gotcha depending on how deeply integrated it is into your system. This means that we, as maintainers, will likely spend a lot of time trying to chase down unintended consequences of changes in embeddings we have limited access to.

I feel that the best compromise between these worlds is regularly expected breaking changes (mild breaking) and users having the expectation that they'll have to bump the version. This creates a number of motivations towards what I feel is "the right thing to do":

* As authors we don't feel hindered landing breaking changes (minor ones)
* As authors we are vigilant to clearly document any possible breakage in release notes
* Users should expect that upgrades are hassle-free, not literally free
* Users should expect to read the release notes and determine risks to their application, if any

I don't think it's viable to have a situation where we, as the Wasmtime project, take on 100% of the work of keeping everyone else's applications running. Our job, in my opinion, is providing a solid WebAssembly runtime that users can rely on. Requiring a small amount of end-user assistance, though, greatly helps actually releasing new features and not feeling hindered in development.